workflow-core/src/main/java/com/squareup/workflow/RenderContext.kt

```diff
@@ -196,12 +185,56 @@ fun <PropsT, StateT, OutputT> RenderContext<PropsT, StateT, OutputT>.runningWork
   worker: Worker<Nothing>,
   key: String = ""
 ) {
-  // Need to cast to Any so the compiler doesn't complain about unreachable code.
-  runningWorker(worker as Worker<Any>, key) {
+  runningWorker(worker, key) {
+    // The compiler thinks this code is unreachable, and it is correct. But we have to pass a lambda
+    // here so we might as well check at runtime as well.
+    @Suppress("UNREACHABLE_CODE", "ThrowableNotThrown")
     throw AssertionError("Worker<Nothing> emitted $it")
   }
 }
 
+/**
+ * Ensures [worker] is running. When the [Worker] emits an output, [handler] is called
+ * to determine the [WorkflowAction] to take. When the worker finishes, nothing happens (although
+ * another render pass may be triggered).
+ *
+ * @param key An optional string key that is used to distinguish between identical [Worker]s.
+ */
+@OptIn(ExperimentalStdlibApi::class)
+/* ktlint-disable parameter-list-wrapping */
+inline fun <T, reified W : Worker<T>, PropsT, StateT, OutputT>
+    RenderContext<PropsT, StateT, OutputT>.runningWorker(
+  worker: W,
+  key: String = "",
+  noinline handler: (T) -> WorkflowAction<PropsT, StateT, OutputT>
+) {
+/* ktlint-enable parameter-list-wrapping */
+  runningWorker(worker, typeOf<W>(), key, handler)
```

Comment:
This isn't the API I'd ideally like to have, but since we don't have access to the full reflection library in this module, it's the best we can do. If we did, we could create the full worker type from the worker's class and the worker's `outputType` property, or maybe using the reified `T` type parameter. But since we don't, all we have is `typeOf()`, which means we need to reify the entire worker type.

However, while less flexible, it has the nice property that the type that is used to compare the worker will _always_ be _exactly_ the type for the worker that the IDE is showing you when you pass it to `runningWorker`. There's no mystery or hidden implementation details anymore. You might need to pass a few more `key`s in, but that's probably a good idea anyway.